### PR TITLE
Created DocClean

### PR DESCRIPTION
**Pull request by devanshkv**
Added scripts for training autoencoders and Cycle GAN

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    version='0.1',
+    packages=find_packages(),
+    install_requires=required,
+    tests_require=['pytest'],
+    zip_safe=False,
+    url='https://github.com/devanshkv/insight_docclean/',
+    license='GNU Lesser General Public License v3.0',
+    author='Devansh Agarwal',
+    author_email='devansh.kv@gmail.com',
+    description='Clean your document images'
```
Comment (michaelliu2) [inline setup.py]: Missing comma. I'd fix that and just make sure `python setup.py install` command works

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+classifiers = [
```
Comment (michaelliu2) [inline setup.py]: Nitpick linting issue: should be `classifier=[`. Also missing indent

Comment (devanshkv) [reply setup.py]: Fixed with commit 4d6d6c3.

File: docclean/autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                 early_stopping: int = 25):
+        self.img_rows = img_rows
+        self.img_cols = img_cols
+        self.channels = nchans
+        self.early_stopping = early_stopping
+        self.img_shape = (self.img_rows, self.img_cols, self.channels)
+
+        if mixed_precision:
+
+            self.policy = mixed_precision.Policy('mixed_float16')
```
Comment (michaelliu2) [inline docclean/autoencoder.py:35]: mixed_precision is a `bool`. Do you need to import this library from tensorflow? Alternatively, if you're not using it, you just could remove it

File: docclean/cycle_gan.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+class CycleGan:
+    """
+    Cycle GAN in tensorflow based on the pip2pix.
+
+    Args:
+
+        checkpoint_path: Where to put the checkpoints
+
+        restor_checpoint: Restore old checkpoints or not
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: Nitpick: maybe spell it `restore_checkpoint` ?

File: docclean/autoencoder.py (hunk 3 above)
Comment (devanshkv) [reply docclean/autoencoder.py:35]: Commit 9767677 fixes this.

File: docclean/cycle_gan.py (hunk 4 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Aah it's a spelling error, on it.

File: docclean/cycle_gan.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        generated_loss = self.loss_obj(tf.zeros_like(generated), generated)
+        total_disc_loss = real_loss + generated_loss
+        return total_disc_loss * 0.5
+
+    def generator_loss(self, generated):
+        return self.loss_obj(tf.ones_like(generated), generated)
+
+    def calc_cycle_loss(self, real_image, cycled_image):
+        loss1 = tf.reduce_mean(tf.abs(real_image - cycled_image))
+        return LAMBDA * loss1
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: `self.LAMBDA`? Same in line 77

File: docclean/cycle_gan.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        self.discriminator_x_optimizer.apply_gradients(zip(discriminator_x_gradients,
+                                                           discriminator_x.trainable_variables))
+
+        self.discriminator_y_optimizer.apply_gradients(zip(discriminator_y_gradients,
+                                                           discriminator_y.trainable_variables))
+
+    def train(self, dirty_images: tf.data.Dataset, clean_images: tf.data.Dataset, epochs: int = 50):
+        for epoch in tqdm.tqdm(range(epochs), leave=False):
+            for image_x, image_y in tqdm.notebook.tqdm(tf.data.Dataset.zip((dirty_images, clean_images)), leave=False):
+                train_step(image_x, image_y)
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: `self.train_step`?

File: docclean/cycle_gan.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+        self.discriminator_x_optimizer.apply_gradients(zip(discriminator_x_gradients,
+                                                           discriminator_x.trainable_variables))
+
+        self.discriminator_y_optimizer.apply_gradients(zip(discriminator_y_gradients,
+                                                           discriminator_y.trainable_variables))
+
+    def train(self, dirty_images: tf.data.Dataset, clean_images: tf.data.Dataset, epochs: int = 50):
+        for epoch in tqdm.tqdm(range(epochs), leave=False):
+            for image_x, image_y in tqdm.notebook.tqdm(tf.data.Dataset.zip((dirty_images, clean_images)), leave=False):
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: do you need `tqdm.notebook`? I guess if you want to run it in a notebook

File: docclean/cycle_gan.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            # Total generator loss = adversarial loss + cycle loss
+            total_gen_g_loss = gen_g_loss + total_cycle_loss + self.identity_loss(real_y, same_y)
+            total_gen_f_loss = gen_f_loss + total_cycle_loss + self.identity_loss(real_x, same_x)
+
+            disc_x_loss = self.discriminator_loss(disc_real_x, disc_fake_x)
+            disc_y_loss = self.discriminator_loss(disc_real_y, disc_fake_y)
+
+        # Calculate the gradients for generator and discriminator
+        generator_g_gradients = tape.gradient(total_gen_g_loss,
+                                              generator_g.trainable_variables)
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: in the following lines, I believe `generator_g`, `generator_f`, `discriminator_x`, `discriminator_y` should all have `self.`

File: docclean/cycle_gan.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
+import tensorflow as tf
+import tqdm
+from tensorflow_examples.models.pix2pix import pix2pix
+
+
+class CycleGan:
+    """
+    Cycle GAN in tensorflow based on the pip2pix.
```
Comment (michaelliu2) [inline docclean/cycle_gan.py]: nitpick: `pix2pix` not `pip2pix`

File: docclean/cycle_gan.py (hunk 5 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Fixed in 92ee6ee.

File: docclean/cycle_gan.py (hunk 6 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Fixed in 92ee6ee.

File: docclean/cycle_gan.py (hunk 7 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Fixed in 92ee6ee.

File: docclean/cycle_gan.py (hunk 8 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Fixed in 92ee6ee.

File: docclean/cycle_gan.py (hunk 9 above)
Comment (devanshkv) [reply docclean/cycle_gan.py]: Fixed in 92ee6ee.

File: train.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+        kaggle_dirty_images = kaggle_dirty_images.map(docclean.utils.kaggle_crop_and_augment,
+                                                      num_parallel_calls=AUTOTUNE)
+        kaggle_clean_images = kaggle_clean_images.map(docclean.utils.kaggle_crop_and_augment,
+                                                      num_parallel_calls=AUTOTUNE)
+
+        if args.dirty_books_dir is not None:
+            books_dirty_images = tf.data.Dataset.from_tensor_slices(
+                glob.glob(f"{args.dirty_books_dir}/*png").shuffle(buffer_size=40960)
+            books_clean_imgaes = tf.data.Dataset.from_tensor_slices(
```
Comment (michaelliu2) [inline train.py]: misspelled `books_clean_images`

File: train.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            books_clean_images = books_clean_imgaes.map(docclean.utils.get_png_data).cache()
+            books_dirty_images = books_dirty_images.map(docclean.utils.books_crop_and_augment,
+                                                        num_parallel_calls=AUTOTUNE)
+            books_clean_images = books_clean_images.map(docclean.utils.books_crop_and_augment,
+                                                        num_parallel_calls=AUTOTUNE)
+
+            dirty_images = tf.data.Dataset.concatenate(kaggle_dirty_images, books_dirty_images)
+            clean_images = tf.data.Dataset.concatenate(kaggle_clean_images, books_clean_images)
+
+            else:
```
Comment (michaelliu2) [inline train.py]: `else`: is too indented

File: train.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                                                      num_parallel_calls=AUTOTUNE).cache()
+
+        kaggle_dirty_images = kaggle_dirty_images.map(docclean.utils.kaggle_crop_and_augment,
+                                                      num_parallel_calls=AUTOTUNE)
+        kaggle_clean_images = kaggle_clean_images.map(docclean.utils.kaggle_crop_and_augment,
+                                                      num_parallel_calls=AUTOTUNE)
+
+        if args.dirty_books_dir is not None:
+            books_dirty_images = tf.data.Dataset.from_tensor_slices(
+                glob.glob(f"{args.dirty_books_dir}/*png").shuffle(buffer_size=40960)
```
Comment (michaelliu2) [inline train.py]: you're missing a `)` after `*png")`

File: train.py (hunk 10 above)
Comment (devanshkv) [reply train.py]: Fixed in 4806eb8.

File: train.py (hunk 11 above)
Comment (devanshkv) [reply train.py]: Fixed in 4806eb8.

File: train.py (hunk 12 above)
Comment (devanshkv) [reply train.py]: Fixed in 4806eb8.

**Comment by michaelliu2** [comment]
Overall, the code looks good to me, and looks like you already addressed my comments lol. 👍. Feel free to merge.

Also, just in general, it's always up to you in the end, what code changes you want to make and merge. TAs just try to give helpful suggestions.

**Comment by devanshkv** [comment]
Thanks a lot! This was very helpful!

**Pull request merged**